src/FsAutoComplete.Core/Commands.fs

```diff
@@ -749,159 +705,283 @@ module Commands =
+        ///
+        /// `Error` iff adjusting ranges failed (including cannot get source) and `errorOnFailureToFixRange`. Otherwise always `Ok`
+        let tryFindReferencesInFile (file: string<LocalPath>, project: FSharpProjectOptions) =
+          async {
+            if dict.ContainsKey file then
+              return Ok()
             else
-              match sourceSpan.IndexOf(symbolText) with
-              | -1 -> ()
-              | n ->
-                if sourceSpan.Length >= n + symbolText.Length then
-                  let startPos = symbolUseRange.Start.IncColumn n
-                  let endPos = symbolUseRange.Start.IncColumn(n + symbolText.Length)
-
-                  let actualUseRange = Range.mkRange symbolUseRange.FileName startPos endPos
-                  symbolUseRanges.Add actualUseRange
+              let! references = findReferencesForSymbolInFile (file, project, symbol)
+
+              let references =
+                if includeDeclarations then
+                  references
+                else
+                  references |> Seq.filter (not << isDeclLocation)
+
+              let references = references |> Seq.toArray
+
+              // Note: this check is important!
+              // otherwise `tryAdjustRanges` tries to get source for files like `AssemblyInfo.fs`
+              //   (which fails -> error if `errorOnFailureToFixRange`)
+              if references |> Array.isEmpty then
+                return Ok()
+              else
+                let ranges = tryAdjustRanges (file, references)
+
+                match ranges with
+                | Error msg when errorOnFailureToFixRange -> return Error msg
+                | Error _ ->
+                  dict.TryAdd(file, references) |> ignore
+                  return Ok()
+                | Ok ranges ->
+                  dict.TryAdd(file, ranges) |> ignore
+                  return Ok()
           }
+          |> Async.Catch
+          |> Async.map (Result.ofChoice >> Result.mapError string >> Result.bind id)
           |> Async.map (fun x ->
             match x with
             | Ok () -> ()
             | Error e ->
-              commandsLogger.info (Log.setMessage "OnFound failed: {errpr}" >> Log.addContextDestructured "error" e))
+              commandsLogger.info (
+                Log.setMessage "tryFindReferencesInFile failed: {error}"
+                >> Log.addContextDestructured "error" e
+              ))
+
+        let iterProject (project: FSharpProjectOptions) =
+          asyncResult {
+            //Enhancement: do in parallel?
+            for file in project.SourceFiles do
+              let file = UMX.tag file
+              do! tryFindReferencesInFile (file, project)
+          }
 
-        let! _ = getSymbolUsesInProjects (symbol, projects, onFound)
+        let iterProjects (projects: FSharpProjectOptions seq) =
+          asyncResult {
+            for project in projects do
+              do! iterProject project
+          }
```

Comment:
Been running this branch a few days on my work project.  Yeah parallel helps a lot here with speed. I did something like:

```fsharp
        let iterProjects (projects: FSharpProjectOptions seq) =
            [
              for project in projects do
                for file in project.SourceFiles do
                  let file = UMX.tag file
                  tryFindReferencesInFile (file, project)
            ]
            |> Async.Parallel
            |> Async.Ignore<unit array>
```

and also changed the `Dictionary` to a `ConcurrentDictionary`.

Reply:
How does it feel from a load perspective? Async.Parallel queues all the items on the threadpool right, so you by default do NUM_CPUS at once? Would it make sense to limit to e.g. NUM_CPUs/2 or something to make sure there's headroom for other operations? What's the delay like in your experience?

QUESTIONS!

Reply:
> How does it feel from a load perspective? Async.Parallel queues all the items on the threadpool right, so you by default do NUM_CPUS at once?

Mostly the big wait time is the type checker so we're throttled by that until type checks complete for the solution. (mostly 1 core)

Otherwise the main part is looping through a bunch of files and adding them to a dictionary. This part parallelizes extremely well. Looping through a lot of files can be slow comparatively.

Reply:
> I did something like:
> ```fsharp
> let iterProjects (projects: FSharpProjectOptions seq) =
>     [
>       for project in projects do
>         for file in project.SourceFiles do
>           let file = UMX.tag file
>           tryFindReferencesInFile (file, project)
>     ]
>     |> Async.Parallel
>     |> Async.Ignore<unit array>
> ```


One disadvantage of this: it `tryFindReferencesInFile` for all files in all projects -- always.  
[Currently](https://github.com/Booksbaum/FsAutoComplete/blob/c60a55283831253ecf065fd97a7f6993e65d2e70/src/FsAutoComplete.Core/Commands.fs#L868-L880) it stops on first error. (usually just for Rename when it cannot correct the range of a symbol). Is there already a `Async.Parallel` that stops on first error?


Though when deciding between these two: Parallel is probably preferred: It can only fail for Rename, not Find References -- which is most likely more often used. And now Adaptive Server loads all files -> not being able to correct range shouldn't happen that often

Reply:
I think the errors I run into are because I'm in a mixed C#/F# project and it tries to do things for cs file files so maybe we need to have more guards around file types.